### PR TITLE
[GH-1021] Submit arrays workflows using Terra/RAWLs

### PR DESCRIPTION
**Pull request by samanehsan**
### Purpose

Ticket: https://broadinstitute.atlassian.net/browse/GH-1021

### Changes

Update the arrays module to start a workflow in the Terra Cromwell, via the firecloud API.

Example for create/exec:
```
curl --location --request POST 'http://localhost:3000/api/v1/create' \
--header 'Authorization: Bearer '$(gcloud auth print-access-token) \
--header 'Content-Type: application/json' \
--data-raw '{
 "cromwell": "https://firecloud-orchestration.dsde-dev.broadinstitute.org",
 "input": "placeholder",
 "output": "placeholder",
 "pipeline": "GPArrays",
 "project": "general-dev-billing-account/arrays",
 "items": [{
     "inputs": { 
        "entity-name": "200598830050_R07C01-1",
        "entity-type": "sample"
       }
 }]}'
```

Notes: 
- We do not have control over workflow labels or options when submitting through terra. This includes setting the final workflow outputs directory. Instead the outputs get written back into the workspace data table.
- This module assumes that the [Arrays workflow](https://bvdp-saturn-dev.appspot.com/#workspaces/general-dev-billing-account/arrays/workflows/general-dev-billing-account/Arrays) and [sample data](https://bvdp-saturn-dev.appspot.com/#workspaces/general-dev-billing-account/arrays/data) are already in the workspace. In the future, WFL could take in the inputs directly and upload the data before running analysis (this is why I left in all of the input-related code).
- This module uses the `cromwell` field for the terra url for now. This should be renamed to `executor` as @tbl3rd suggested.

### Review Instructions

- No instructions.

### CODE REVIEW

File: api/src/wfl/service/terra.clj [hunk 1]
```diff
@@ -39,3 +41,11 @@
...
     (util/parse-json (:body response))))
 
+(defn get-workflow-status-by-entity
+  "Get workflow status given a Terra submission-id and entity-name."
+  [terra-url workspace {:keys [uuid inputs] :as _item}]
+  (->> (get-submission terra-url workspace uuid)
+       :workflows
+       (filter #(= (:entity-name inputs) (get-in % [:workflowEntity :entityName])))
+       (first)
+       (:status)))
```
Comment (ehigham) [inline api/src/wfl/service/terra.clj:51]: What happens if the submission is not found? does get-submission throw or will this return null?

File: api/src/wfl/service/postgres.clj [hunk 2]
```diff
@@ -71,6 +72,18 @@
...
+  [tx {:keys [cromwell items workflows project] :as _workload}]
+  (letfn [(update! [{:keys [id uuid]} status]
+              (jdbc/update! tx items
+                            {:updated (OffsetDateTime/now) :uuid uuid :status status}
+                            ["id = ?" id]))]
+      (->> workflows
+           (remove (comp nil? :uuid))
+           (remove (comp finished? :status))
+           (run! #(update! % (terra/get-workflow-status-by-entity cromwell project %))))))
+
```
Comment (ehigham) [inline api/src/wfl/service/postgres.clj:86]: This code is really similar to the one for cromwell - I wonder if we can re-use some of the logic by creating a higher-order function:

```clojure
(defn ^:private make-update-workflows [get-status!]
  (fn [tx {:keys [items workflows] :as workload}]
    (letfn [(update! [{:keys [id uuid]} status]
              (jdbc/update! tx items
                  {:updated (OffsetDateTime/now) :uuid uuid :status status}
                  ["id = ?" id]))]
       (->> workflows
         (remove (comp nil? :uuid))
         (remove (comp finished? :status))
         (run! #(update! % (get-status! workload %)))))))

(def update-workflow-statuses!
  (letfn [(get-cromwell-status [{:keys [cromwell]} {:keys [uuid]}]
            (if (util/uuid-nil? uuid) "skipped"
                                      (cromwell-status cromwell uuid)))]
    (make-update-workflows get-cromwell-status)))

(def update-terra-workflow-statuses!
  (letfn [(get-terra-status [{:keys [cromwell project]} workflow]
            (terra/get-workflow-status-by-entity cromwell project workflow))]
    (make-update-workflows get-terra-status)))
```

**Review by ehigham — APPROVED** [review]
Looks good!

**Comment by ehigham** [comment]
Will you add tests to lock down behaviour?

File: api/src/wfl/service/terra.clj (hunk 1 above)
Comment (samanehsan) [reply api/src/wfl/service/terra.clj:51]: In that case `get-submission` would return a 404 error

File: api/src/wfl/service/postgres.clj [hunk 3]
```diff
@@ -55,21 +56,29 @@
...
+    (letfn [(update! [{:keys [id uuid]} status]
+              (jdbc/update! tx items
+                            {:updated (OffsetDateTime/now) :uuid uuid :status status}
+                            ["id = ?" id]))]
+      (->> workflows
+           (remove (comp nil? :uuid))
+           (remove (comp finished? :status))
+           (run! #(update! % (get-status! workload %)))))))
+
+(def update-workflow-statuses!
```
Comment (ehigham) [inline api/src/wfl/service/postgres.clj:71]: Can you move the doc string from the private function  (`make-update-workflows`) onto these two new implementations? Don't think the private function has much business having a doc string....

File: api/src/wfl/service/postgres.clj [hunk 4]
```diff
@@ -55,21 +56,29 @@
   "Test if a workflow `:status` is in a terminal state."
   (set (conj cromwell/final-statuses "skipped")))
 
-(defn update-workflow-statuses!
+(defn ^:private make-update-workflows [get-status!]
   "Use `tx` to update `status` of `workflows` in `_workload`."
```
Comment (ehigham) [inline api/src/wfl/service/postgres.clj:60]: this doc string is out-dated

File: api/src/wfl/service/terra.clj (hunk 1 above)
Comment (ehigham) [reply api/src/wfl/service/terra.clj:51]: Thanks!

**Review by ehigham — APPROVED** [review]

File: api/src/wfl/service/postgres.clj (hunk 3 above)
Comment (samanehsan) [reply api/src/wfl/service/postgres.clj:71]: Good point!

**Pull request merged**